--- a/libdropbear.h	2019-07-11 11:45:13.848000000 +0530
+++ b/libdropbear.h	2019-07-12 17:36:05.210348598 +0530
@@ -0,0 +1,28 @@
+#ifndef __LIBDROPBEAR_H
+#define __LIBDROPBEAR_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define	IPADDRESSLENGTH		16
+#define	TIMESTAMPLENGTH		26
+
+struct client_info{
+        int  pid;
+        char ipaddress[IPADDRESSLENGTH];  //Client IPv4 address
+        char timestamp[TIMESTAMPLENGTH];  //timestamp of the received client
+};
+
+void activate_dropbear(char* , char* , char*);
+void deactivate_dropbear();
+int get_active_sessions_count();
+void get_active_sessions_info(struct client_info*, int);
+int close_client_session(int);
+int get_total_sessions_served();
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
